### PR TITLE
CAMEL-16674 - Base commit for huawei cloud function graph component

### PR DESCRIPTION
**Pull request by hokutor**
JIRA - https://issues.apache.org/jira/browse/CAMEL-16674

FunctionGraph is Huawei cloud's offering for serverless computation service that lets you run code without provisioning or managing servers, creating workload-aware cluster scaling logic, maintaining event integrations, or managing runtimes.  This service allows you to run your code ensuring high availability and scalability. All you need to do is upload your code and set execution conditions and FunctionGraph will take care of the rest. Refer https://www.huaweicloud.com/intl/en-us/product/functiongraph.html for more information

This PR contains camel component for integrating with already existing functions (created by the cloud user via huaweicloud webconsole)

### CODE REVIEW

**Comment by hokutor** [comment]
Why is the mvn sourcecheck failing on camel-debezium-common component? Is there anything we can do about this?

**Comment by davsclaus** [comment]
Ah its not checkstyle its java 8 compilation error it seems after upgrading DBZ from 1.5.0 to 1.5.1.

**Comment by davsclaus** [comment]
Okay 1.5.1 had a java 8 bug its fixed by the new 1.5.2

**Comment by davsclaus** [comment]
The pom.xml file should be updated to be like the other pom files we have for components out of the box (it looks like it was standalone), and it should also have license headers

File: components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        if (ObjectHelper.isNotEmpty(clientConfigurations.getXCffLogType())) {
+            exchange.setProperty(FunctionGraphProperties.XCFFLOGS, response.getLog());
+        }
+        LOG.info("Invoke Function results: " + response);
+    }
+
+    /**
+     * Initialize the client
+     */
+    private void initClient() {
```
Comment (oscerd) [inline components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java:131]: This should be done in the endpoint. I don't think you'll change the client options many times. So it makes sense to do this just one time in endpoint and use client instance in producer.

**Review by oscerd — CHANGES_REQUESTED** [review]
The component completely miss the needed steps to be added to kit (references in parent POM, components POM and BOM). There are other point missing and some errors. But it is a good start. Thanks.

File: components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        Map request;
+        if (body == null) {
+            if (LOG.isErrorEnabled()) {
+                LOG.error("Exchange body is null");
+            }
+            throw new IllegalArgumentException("Exchange body is mandatory and should be a valid Map or JSON string");
+        } else if (body instanceof Map) {
+            request = exchange.getMessage().getBody(Map.class);
+        } else {
+            String strBody = exchange.getMessage().getBody(String.class);
```
Comment (omarsmak) [inline components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java:84]: Here you may face the case where the body is not a `String` nor `Map`

File: components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    public FunctionGraphProducer(FunctionGraphEndpoint endpoint) {
+        super(endpoint);
+        this.endpoint = endpoint;
+    }
+
+    public void process(Exchange exchange) throws Exception {
+        updateClientConfigs(exchange);
+
+        switch (clientConfigurations.getOperation()) {
+            case FunctionGraphOperations.INVOKE_FUNCTION:
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java:63]: We will be supporting more operations soon

File: components/camel-huaweicloud-functiongraph/pom.xml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  <version>3.11.0-SNAPSHOT</version>
+
+  <name>Camel FunctionGraph Component</name>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+  </properties>
+
+  <dependencyManagement>
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/pom.xml]: Done. Using components parent pom. created a module entry in components pom for functiongraph.

File: components/camel-huaweicloud-functiongraph/pom.xml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    <dependency>
+      <groupId>org.apache.camel</groupId>
+      <artifactId>camel-support</artifactId>
+    </dependency>
+
+    <!-- HuaweiCloud SDK v3 -->
+    <dependency>
+      <groupId>com.huaweicloud.sdk</groupId>
+      <artifactId>huaweicloud-sdk-functiongraph</artifactId>
+      <version>3.0.43-rc</version>
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/pom.xml]: Done

File: components/camel-huaweicloud-functiongraph/pom.xml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    </dependency>
+
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+      <version>2.8.6</version>
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/pom.xml]: Done

File: components/camel-huaweicloud-functiongraph/pom.xml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      <artifactId>log4j-slf4j-impl</artifactId>
+      <version>2.13.3</version>
+      <scope>test</scope>
+    </dependency>
+
+    <!-- testing -->
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <version>5.7.0</version>
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/pom.xml]: Done

File: components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java (hunk 1 above)
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java]: Done

File: components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java (hunk 2 above)
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphProducer.java:75]: Handled

**Comment by hokutor** [comment]
After integrating with the kit, we get following error while building the component. I've made sure the [firstVersion is specified](https://github.com/hokutor/camel/blob/CAMEL-16674/components/camel-huaweicloud-functiongraph/src/main/java/org/apache/camel/FunctionGraphEndpoint.java#L34) in the endpoint class. Can you suggest on what couldn't be wrong here?

```logs
[INFO] No previous run data found, generating files.
[INFO] Building index...
[INFO] Updated org/apache/camel/FunctionGraphEndpointUriFactory.java
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10:27 min
[INFO] Finished at: 2021-06-02T11:01:23-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.camel:camel-package-maven-plugin:3.11.0-SNAPSHOT:generate (generate) on project camel-huaweicloud-functiongraph: Error generating data org.apache.maven.plugin.MojoExecutionException: Error l
oading other model. Reason: java.lang.IllegalArgumentException: FirstVersion is not specified. This can be done in @UriEndpoint or in pom.xml file. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

File: components/camel-huaweicloud-functiongraph/pom.xml [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+      <version>2.8.6</version>
+    </dependency>
+
+    <!-- logging -->
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-slf4j-impl</artifactId>
+      <version>2.13.3</version>
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/pom.xml]: Done

File: components/camel-huaweicloud-functiongraph/pom.xml [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+  <build>
+    <defaultGoal>install</defaultGoal>
+
+    <plugins>
+      
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.8.1</version>
```
[the remark this answers is not in the document]
Comment (hokutor) [reply components/camel-huaweicloud-functiongraph/pom.xml]: Done

**Review by oscerd — CHANGES_REQUESTED** [review]
The component is missing in parent/pom and in camel-bom/pom

**Comment by hokutor** [comment]
> The component is missing in parent/pom and in camel-bom/pom

Thanks @oscerd for bringing that up. Added to parent/pom and camel-bom/pom

**Review by oscerd — APPROVED** [review]

**Comment by davsclaus** [comment]
Would it be possible to make configuring those secret and auth keys (and the likes) on the component level as well. Then you can configure this once, and also more easily via Camels support for this, so not everything is on the endpoint.

You can keep the option on the endpoint level as well, so you can choose.

**Comment by hokutor** [comment]
> Would it be possible to make configuring those secret and auth keys (and the likes) on the component level as well. Then you can configure this once, and also more easily via Camels support for this, so not everything is on the endpoint.
> 
> You can keep the option on the endpoint level as well, so you can choose.

Thanks Claus for the suggestion. Can you give us a reference ? Is there another component which we can refer to understand how it is to be done well?

**Comment by davsclaus** [comment]
> > Would it be possible to make configuring those secret and auth keys (and the likes) on the component level as well. Then you can configure this once, and also more easily via Camels support for this, so not everything is on the endpoint.
> > You can keep the option on the endpoint level as well, so you can choose.
> 
> Thanks @davsclaus for the suggestion. Can you give us a reference ? Is there another component which we can refer to understand how it is to be done well?

We can do this later in a new PR. I suggest to polish and make this PR ready for review and merge.

**Comment by hokutor** [comment]
> > > Would it be possible to make configuring those secret and auth keys (and the likes) on the component level as well. Then you can configure this once, and also more easily via Camels support for this, so not everything is on the endpoint.
> > > You can keep the option on the endpoint level as well, so you can choose.
> > 
> > 
> > Thanks @davsclaus for the suggestion. Can you give us a reference ? Is there another component which we can refer to understand how it is to be done well?
> 
> We can do this later in a new PR. I suggest to polish and make this PR ready for review and merge.

Perfect. Thanks @davsclaus . Let me add one more functional test case. and submit for review and merge

**Comment by hokutor** [comment]
@davsclaus should be ready for review and merge now

**Review by davsclaus — APPROVED** [review]

**Pull request merged**